### DISCUSSION TITLE
Details: the method eth_sendTransaction is not available Version: viem@1.18.1

### DISCUSSION THREAD

**Discussion by 0xwagyu-eth** [wevm/viem#1452] [Question] [answered]
Coming across the below error when trying to write to contract. 

 Details: the method eth_sendTransaction is not available Version: viem@1.18.1

```
export async function rebalance(
  publicClient: PublicClient,
  walletClient: WalletClient,
  address: 0x${string},
) {

  const { request } = await publicClient.simulateContract({
    address: address,
    abi: Enigma_ABI,
    functionName: "rebalance",
    args: [someargs],
  });

  await walletClient.writeContract(request);

}
```

**Comment by jxom** [comment]
What transport are you using for your wallet client? You'll need to make sure that the transport supports the `eth_sendTransaction` method (I suspect you may be trying to send a transaction on a public RPC/node like Alchemy, Infura, etc, and not via a wallet).

**Reply by 0xwagyu-eth** [reply]
just using http() as transport. I need to be able to write() without having to approve from a wallet. This will be a simple private nodeJS environment without any frontend. Is there a suggested approach for this?

**Comment by joaofranciscoguarda** [comment]
I had the same error, instead you had to use `sendRawTransaction`, but for this, you need to use the method sendRawTransaction, passing the sign transaction to `serializedTrasaction` property. 

I suppose you are using an ankr RPC, using Infura you recieve a sendTransaction is not supported/implemented, so the problem is not related to the RPC, and is about the method you are using to send. 

Viem should have a alternative for the writeContract using ONLY sendTransaction. And I believe this happens because wagmi is mostly used in the front, but viem is rising in the nodeJs enviroment, specially in the back.

**Reply by 0xwagyu-eth** [reply]
Yes, I am building a private nodeJS environment without any front facing ui. 

Did you have any examples you would be able to share with me on this? I feel like this should be easier as it would be common use case

**Reply by joaofranciscoguarda** [reply]
Sorry for the late return. 

Well, I used this:
```typescript
const newTx = await this.client.sendRawTransaction({ serializedTransaction: signedTx });
```

But for this, it's need to have the private key saved in some place, I'm using a database
```typescript
// Get the private key
const pk = wallet.getData<Hash>("pk");
// Create the account using the pk
const account = privateKeyToAccount(pk);
// Signs the Tx
const signedTx = await account.signTransaction({
      ...calldata, 
      nonce: Number(nonce) // This guy may be in bigint, number is required
});

return  signedTx
```

Having the pk or not is another discussion, but for the begging this is something to start with.
Have a good one!

**Comment by mikefuntico** [comment, marked as answer]
You need to pass `account` to writeContract derived as:
```js
import { privateKeyToAccount } from 'viem/accounts'


const account = privateKeyToAccount(PRIVATE_KEY)

const hash = await client.writeContract({
    abi,
    address: contractAddress,
    functionName: 'rebalance',
    args: [someargs],
    account,
  })
```
and not just an address. After that it will work as intended.

**Reply by fadeevab** [reply]
If you're using `simulateContract`, ensure you use the same account which holds a private key:

```ts
const account = privateKeyToAccount(privateKey); // type: 'local'
const client = createPublicClient({ chain, transport });
const wallet = createWalletClient({ chain, transport, account }); // <--- private key local account

const { request } = await client.simulateContract({
    account: wallet.account, // <--- !!! the same private key local account
    address: contract,
    abi: ATLAS_ABI,
    functionName: 'withdrawSurcharge',
});

const hash = await this.wallet.writeContract(request); // <--- now, request's account == wallet.account
```
